src/cfi.rs

```diff
@@ -89,6 +90,149 @@ impl<R: Reader> From<R> for DebugFrame<R> {
+        let table_enc = DwEhPe(reader.read_u8()?);
+        let eh_frame_ptr = parse_encoded_pointer(eh_frame_ptr_enc, bases, addr_size, &self.0, &mut reader)?;
+        let fde_count = parse_encoded_pointer(fde_count_enc, bases, addr_size, &self.0, &mut reader)?;
+        let fde_count = match fde_count {
+            Pointer::Direct(c) => c,
+            Pointer::Indirect(_) => unreachable!(),
+        };
+
+        Ok(ParsedEhFrameHdr {
+            addr_size,
+            section: self.0.clone(),
+
+            eh_frame_ptr,
+            fde_count,
+            table_enc,
+            table: reader,
+        })
+    }
+}
+
+impl<R: Reader> Section<R> for EhFrameHdr<R> {
+    fn section_name() -> &'static str {
+        ".eh_frame_hdr"
+    }
+}
+
+impl<R: Reader> From<R> for EhFrameHdr<R> {
+    fn from(section: R) -> Self {
+        EhFrameHdr(section)
+    }
+}
+
+impl<R: Reader> ParsedEhFrameHdr<R> {
+    /// Returns the address of the binary's `.eh_frame` section.
+    pub fn eh_frame_ptr(&self) -> Pointer {
+        self.eh_frame_ptr
+    }
+
+    /// Retrieves the CFI binary search table, if there is one.
+    pub fn table(&self) -> Option<EhHdrTable<R>> {
+        if self.table_enc == constants::DW_EH_PE_omit {
+            None
+        } else {
+            Some(EhHdrTable { hdr: self })
+        }
+    }
+}
+
+/// The CFI binary search table that is an optional part of the `.eh_frame_hdr` section.
+#[derive(Debug, Clone)]
+pub struct EhHdrTable<'a, R: Reader + 'a> {
+    hdr: &'a ParsedEhFrameHdr<R>,
+}
+
+impl<'a, R: Reader + 'a> EhHdrTable<'a, R> {
+    /// Returns a pointer to the FDE for the given address.
+    ///
+    /// This performs a binary search.
+    pub fn lookup(&self, address: u64, bases: &BaseAddresses) -> Result<Pointer> {
+        let size = match self.hdr.table_enc.format() {
+            constants::DW_EH_PE_uleb128 | constants::DW_EH_PE_sleb128
+                => return Err(Error::VariableLengthSearchTable),
+            constants::DW_EH_PE_sdata2 | constants::DW_EH_PE_udata2 => 2,
+            constants::DW_EH_PE_sdata4 | constants::DW_EH_PE_udata4 => 4,
+            constants::DW_EH_PE_sdata8 | constants::DW_EH_PE_udata8 => 8,
+            _ => return Err(Error::UnknownPointerEncoding),
+        };
+
+        let row_size = size * 2;
+
+        let mut len = self.hdr.fde_count;
+
+        let mut reader = self.hdr.table.clone();
+
+        while len > 1 {
+            let head = reader.split(R::Offset::from_u64((len / 2) * row_size)?)?;
+            let tail = reader.clone();
+
+            let pivot = parse_encoded_pointer(self.hdr.table_enc, bases, self.hdr.addr_size,
+                                              &self.hdr.section, &mut reader)?;
+            let pivot = match pivot {
+                Pointer::Direct(x) => x,
+                Pointer::Indirect(_) => return Err(Error::UnsupportedPointerEncoding),
+            };
+
+            match pivot.cmp(&address) {
```

Comment:
I guess the reason that we can't read the table into a slice and use `slice::binary_search_by` is because we would need to repeat that code for all the different sizes?

I'd prefer to use the off-the-shelf binary search implementation if possible.

Reply:
Perhaps some macro_rules to hide the code repetitition?

Reply:
I actually originally intended to use std's binary search here. I ended up doing it this way instead when I didn't find a safe way (as in: not `transmute`) to convert from `&[u8]` to `&[u32]` and friends.

Reply:
Yeah, I guess that `transmute` involves making sure that alignment is correct and all that too or else we'll perform unaligned loads and go into UB.

Ok, let's keep this binary search implementation.